tests/performance/performance_test_driver.h

```diff
@@ -0,0 +1,120 @@
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_performance_test_driver_h
+#define dealii_performance_test_driver_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+#include <string>
+#include <vector>
+
+
+/**
+ * An enum class that describes the environment in which performance tests
+ * are running.
+ */
+enum class TestingEnvironment
+{
+  /**
+   * A lightweight mobile laptop with at least 2 physical cores and at
+   * least 8 GB of main memory.
+   */
+  light,
+  /**
+   * A medium sized work station with at least 8 physical cores and at
+   * least 32 GB of main memory.
+   */
+  medium,
+  /**
+   * A heavyweight compute node with at least 32 physical cores and at
+   * least 128 GB of main memory.
+   */
+  heavy,
+};
+
+
+/**
+ * Returns a list of timer descriptions.
+ */
+std::vector<std::string>
+describe_benchmarks(const TestingEnvironment &);
+
+/**
+ * Perform a single benchmark run returning a list of timing results (in
+ * seconds).
+ */
+std::vector<double>
+perform_single_benchmark_run(const TestingEnvironment &);
+
+
+int main()
+{
+#ifndef ENABLE_PERFORMANCE_TESTS
+  std::cout << "Performance test framework disabled, do not run test."
+            << std::endl;
+  return 1;
+#endif
+
+  // FIXME: will be supplied via preprocessor macro
+  const TestingEnvironment environment = TestingEnvironment::light;
+
+  // a small helper macro to print a tab-separated vector of strings.
+  const auto print_string_vector = [](const auto vec) {
+    std::cout << std::accumulate(vec.begin(),
+                                 vec.end(),
+                                 std::string(),
+                                 [](auto a, auto b) {
+                                   if (a.empty())
+                                     return b;
+                                   else
+                                     return a + "\t" + b;
+                                 })
+              << "\n";
+  };
+
+  // Print a tab-separated list of timer names:
+
+  const auto description = describe_benchmarks(environment);
+  print_string_vector(description);
+
+  // For the time being we implement a simple best of two strategy:
+
+  const auto run_1 = perform_single_benchmark_run(environment);
+  Assert(description.size() == run_1.size(), ExcInternalError());
+
+  const auto run_2 = perform_single_benchmark_run(environment);
+  Assert(description.size() == run_2.size(), ExcInternalError());
+
+  std::vector<std::string> result;
+  std::transform(run_1.begin(),
+                 run_1.end(),
+                 run_2.begin(),
+                 std::back_inserter(result),
+                 [](auto left, auto right) {
+                   const auto minimum = std::min(left, right);
+                   return std::to_string(minimum);
+                 });
```

Comment:
I suggest we run a best of two benchmark test for the time being and adjust as needed.